File: old_implementations/Data Structures/Stack/Stack.js
// Stack implementation in JavaScript using objects as storage. Functions include:
// - Push
// - Pop
// - Peek
// - isEmpty

var Stack = function () {
	var top = null;

	this.push = function(element) {
		if(!element)
			return false;

		var node = {
			data: element,
			next: null
		}

		if(!top)
			top = node;
		else {
			node.next = top;
			top = node;
		}
		return true;
	}

	this.pop = function() {
		if(!top)
			return null;
		var ret = top.data;
		top = top.next;
		return ret;
	}

	this.peek = function() {
		if(!top)
			return null;
		return top.data;
	}

	this.isEmpty = function() {
		if(!top)
			return true;
		return false;
	}
}

console.log("Create new stack");
var stack = new Stack();
console.log("Push 7 into stack");
console.log(stack.push(7));
console.log("Push null into stack");
console.log(stack.push(null));
console.log("Push 3 into stack");
console.log(stack.push(3));
console.log("Peek");
console.log(stack.peek());
console.log("is empty");
console.log(stack.isEmpty());
console.log("Pop");
console.log(stack.pop());
console.log("Peek");
console.log(stack.peek());
console.log("Pop");
console.log(stack.pop());
console.log("Peek");
console.log(stack.peek());
console.log("is empty");
console.log(stack.isEmpty());

File: python/data_structures/stack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:
    def __init__(self):
        self.head = None

    def insert(self, value):
        if self.head is None:
            self.head = Node(value)
        else:
            new_head = Node(value)
            new_head.next = self.head
            self.head = new_head

    def pop(self):
        if self.head is None:
            return
        return_value = self.head.value
        self.head = self.head.next
        return return_value


stack = Stack()
stack.insert(1)
stack.insert(2)
stack.insert(3)
stack.insert(4)

assert stack.pop() == 4
assert stack.pop() == 3
stack.insert(3)
assert stack.pop() == 3
assert stack.pop() == 2
assert stack.pop() == 1
